spyder/plugins/ipythonconsole/widgets/main_widget.py

```diff
@@ -650,6 +675,90 @@ def update_actions(self):
                 self.syspath_action.setEnabled(not error_or_loading)
                 self.show_time_action.setEnabled(not error_or_loading)
 
+    def _get_envs(self):
+        """Get the list of environments in the system."""
+        # Compute info of default interpreter to have it available in
+        # case we need to switch to it. This will avoid lags when
+        # doing that in get_value.
+        #if self.default_interpreter not in self.path_to_env:
+        #    self._get_env_info(self.default_interpreter)
+
+        # Get envs
+        conda_env = get_list_conda_envs()
+        pyenv_env = get_list_pyenv_envs()
+        return {**conda_env, **pyenv_env}
+
+    def _get_env_info(self, path):
+        """Get environment information."""
+        path = path.lower() if os.name == 'nt' else path
+        try:
+            name = self.path_to_env[path]
+        except KeyError:
+            if (
+                self.default_interpreter == path
+                and (running_in_mac_app() or is_pynsist())
+            ):
+                name = 'internal'
+            elif 'conda' in path:
+                name = 'conda'
+            elif 'pyenv' in path:
+                name = 'pyenv'
+            else:
+                name = 'custom'
+            version = get_interpreter_info(path)
+            self.path_to_env[path] = name
+            self.envs[name] = (path, version)
+        __, version = self.envs[name]
+        return f'{name} ({version})'
+
+    def get_envs(self):
+        """
+        Get the list of environments in a thread to keep them up to
+        date.
+        """
+        self._worker_manager.terminate_all()
+        worker = self._worker_manager.create_python_worker(self._get_envs)
+        worker.sig_finished.connect(self.update_envs)
+        worker.start()
+
+    def update_envs(self, worker, output, error):
+        """Update the list of environments in the system."""
+        self.envs.update(**output)
+        for env in list(self.envs.keys()):
+            path, version = self.envs[env]
+            # Save paths in lowercase on Windows to avoid issues with
+            # capitalization.
+            path = path.lower() if os.name == 'nt' else path
+            self.path_to_env[path] = env
+
+    def update_environment_menu(self):
+        """Update context menu information."""
```

Comment:
Probably this should mention that is the submenu related with consoles launched for a specific interpreter. So something like:
```suggestion
        """Update context menu entries to select specific interpreter to launch a console."""
```